docs/Testing.fsx

```diff
@@ -947,4 +956,174 @@ qHisto
+let (preData1,preData2) :float[][] * float [][]=  
+    df
+    |> Frame.getRows
+    |> Series.values
+    |> Seq.map (Series.values >> Seq.toArray >> Array.chunkBySize 3 >> fun x -> x.[0],x.[1])
+    |> Array.ofSeq
+    |> Array.unzip
+
+// After chunking, the datasets are separated by factor and the rowkey is added to the data for later identification.
+let tupel a b = (a,b)
+let data1 = Array.map2 tupel rowheader preData1 
+let data2 = Array.map2 tupel rowheader preData2 
+(**
+Optional: Data can be normalised by median centering using the following function:
+*)
+open SAM
+let corrected1 = 
+    let medCorrect = 
+        medianCentering data1 
+    Array.map2 tupel rowheader medCorrect
+let corrected2 = 
+    let medCorrect = 
+        medianCentering data2
+    Array.map2 tupel rowheader medCorrect
+ 
+(**
+This calculates the median of a column and subtracts it from each value in this column. 
+With data1 and data2 , or corrected1 and corrected2, SAM can be performed.
+*)
+let res = FSharp.Stats.Testing.SAM.twoClassUnpaired 100 0.05 data1 data2 (System.Random(1234))
+(**
+The parameters are number of permutations, FDR, the datasets and a seed. For balanced permutations as in SAM, it seems to be sufficient to use ~ 100 permutations (also default in samR). The seed is used for randomization of the permutations (System.Random()), or can be fixed to achieve the same results multiple times (e.g. System.Random(1234)).
+Result contains now the following information: s0, pi0, delta, upper Cut, lower Cut, positive significant bioitems, negative significant bioitems, non significant bioitems, False Discovery Rate (FDR), and median False Positives.
+The following code will help to achieve the typical SAM plot.
+*)
+
+let SAMChart = 
+
+    let observed = [| res.NegSigBioitem; res.NonSigBioitem; res.PosSigBioitem|] |> Array.concat 
+    let obs = observed |> Array.map (fun x -> x.Statistics) 
+    let expected = res.AveragePermutations |> Array.map (fun x -> x.Statistics)
+    let minDi = Seq.min obs
+    let maxDi = Seq.max obs
+
+
+    // positive significant changes 
+    let posExpected = expected.[res.NegSigBioitem.Length + res.NonSigBioitem.Length .. res.NegSigBioitem.Length + res.NonSigBioitem.Length-1 + res.PosSigBioitem.Length]
+    let posChart = 
+        Chart.Point(posExpected,res.PosSigBioitem |> Array.map (fun x -> x.Statistics))
+        |> Chart.withLineStyle(Color=Color.fromKeyword Green)
+        |> Chart.withTraceInfo("positive change",Visible = Visible.True )
+
+
+    // no significant changes
+    let nonex = expected.[res.NegSigBioitem.Length-1 .. res.NegSigBioitem.Length + res.NonSigBioitem.Length]
+    let nonchart = 
+        Chart.Point(nonex,res.NonSigBioitem |> Array.map (fun x -> x.Statistics))
+        |> Chart.withLineStyle(Color=Color.fromKeyword Gray)
+        |> Chart.withTraceInfo("no change",Visible = Visible.True)
+
+    // negative significant changes 
+    let negex = expected.[0 .. res.NegSigBioitem.Length-1]
```

Comment:
```fsharp
let nonex = expected.[res.NegSigBioitem.Length-1 ..
let negex = expected.[0 .. res.NegSigBioitem.Length-1]
```
Here the (res.NegSigBioitem.Length-1)<sup>th</sup> element is assigned to both, nonex and negex. I think an index error occurred.

Reply:
I think it should be:
```fsharp
let posExpected = expected.[res.NegSigBioitem.Length + res.NonSigBioitem.Length-1 .. res.NegSigBioitem.Length + res.NonSigBioitem.Length + res.PosSigBioitem.Length-1]
let nonex = expected.[res.NegSigBioitem.Length .. res.NegSigBioitem.Length + res.NonSigBioitem.Length-1]
let negex = expected.[0 .. res.NegSigBioitem.Length-1]
```

Please doublecheck